pkg/operator/driverfeaturescontroller.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/openshift/vmware-vsphere-csi-driver-operator/pkg/operator/utils"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/kubernetes"
+	corelister "k8s.io/client-go/listers/core/v1"
+	"strings"
+	"time"
+)
+
+type DriverFeaturesController struct {
+	name                   string
+	targetNamespace        string
+	manifest               []byte
+	kubeClient             kubernetes.Interface
+	operatorClient         v1helpers.OperatorClient
+	configMapLister        corelister.ConfigMapLister
+	infraLister            infralister.InfrastructureLister
+	clusterCSIDriverLister clustercsidriverlister.ClusterCSIDriverLister
+}
+
+func NewDriverFeaturesController(
+	name string,
+	namespace string,
+	manifest []byte,
+	kubeClient kubernetes.Interface,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	operatorClient v1helpers.OperatorClient,
+	configInformer configinformers.SharedInformerFactory,
+	clusterCSIDriverInformer clustercsidriverinformer.ClusterCSIDriverInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	configMapInformer := kubeInformers.InformersFor(namespace).Core().V1().ConfigMaps()
+	infraInformer := configInformer.Config().V1().Infrastructures()
+	c := &DriverFeaturesController{
+		name:                   name,
+		targetNamespace:        namespace,
+		manifest:               manifest,
+		kubeClient:             kubeClient,
+		configMapLister:        configMapInformer.Lister(),
+		operatorClient:         operatorClient,
+		infraLister:            infraInformer.Lister(),
+		clusterCSIDriverLister: clusterCSIDriverInformer.Lister(),
+	}
+	return factory.New().WithInformers(
+		configMapInformer.Informer(),
+		infraInformer.Informer(),
+		clusterCSIDriverInformer.Informer(),
+	).WithSync(
+		c.Sync,
+	).ResyncEvery(
+		time.Minute,
+	).WithSyncDegradedOnError(
+		operatorClient,
+	).ToController(
+		c.name,
+		recorder.WithComponentSuffix("feature-config-controlller-"+strings.ToLower(name)),
+	)
+}
+
+func (d DriverFeaturesController) Sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	opSpec, _, _, err := d.operatorClient.GetOperatorState()
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if opSpec.ManagementState != opv1.Managed {
+		return nil
+	}
+
+	// We will eventually need infra object to get the cluster defaults
+	// TODO: Read the defaults from infra object.
+
+	clusterCSIDriver, err := d.clusterCSIDriverLister.Get(utils.VSphereDriverName)
+	if err != nil {
+		return err
+	}
+
+	defaultFeatureConfigMap := resourceread.ReadConfigMapV1OrDie(d.manifest)
+
+	driverConfig := clusterCSIDriver.Spec.DriverConfig
+	emptyDriverConfig := opv1.CSIDriverConfigSpec{}
+	if driverConfig != emptyDriverConfig {
+		vsphereConfig := driverConfig.VSphere
+		if vsphereConfig != nil && len(vsphereConfig.TopologyCategories) > 0 {
```

Comment:
Driver config spec is a discriminated union, you need a switch here on the discriminant of the union rather than comparing to an empty config.

If you only care about one entry, then you don't need to, but normally what you'd expect is

```
switch driverConfig.DriverType{
case opv1.VSphereDriverType:
  // Do vSphere things here based on if the vSphere config was provided or not
default:
 // My controller /doesdoesn't understand this case, do the appropriate thing
}
```

Reply:
fixed